src/mscorlib/src/System/Reflection/Emit/DynamicILGenerator.cs

```diff
@@ -382,19 +382,20 @@ public override void Emit(OpCode opcode, SignatureHelper signature)
         // Exception related generation
         //
         //
-        public override Label BeginExceptionBlock()
+        public override void BeginExceptFilterBlock()
         {
-            return base.BeginExceptionBlock();
-        }
+            // Begins an exception filter block. Emits a branch instruction to the end of the current exception block.
 
-        public override void EndExceptionBlock()
-        {
-            base.EndExceptionBlock();
-        }
+            if (CurrExcStackCount == 0)
+                throw new NotSupportedException(Environment.GetResourceString("Argument_NotInExceptionBlock"));
 
-        public override void BeginExceptFilterBlock()
-        {
-            throw new NotSupportedException(Environment.GetResourceString("InvalidOperation_NotAllowedInDynamicMethod"));
+            __ExceptionInfo current = CurrExcStack[CurrExcStackCount - 1];
+
+            Label endLabel = current.GetEndLabel();
+            Emit(OpCodes.Leave, endLabel);
+            UpdateStackSize(OpCodes.Nop, 1);
```

Comment:
Notably, the same is true of `BeginCatchBlock` for the non-filter branch, which was the clue to me as to why just removing the exception-throw and letting the base take over resulted in incorrect methods being produced. That seems like a bug in the base too just looking at the code, but it presumably works (`BeginCatchBlock` isn't something super-rare where bugs could hide for ages).